Make avoid overlap work with more digitizing tools (fixes #50433)

JuhoErvasti opened the pull request:
## Description

Use the QgsAvoidIntersectionsOperation class to remove overlap from geometries which are edited with the move, rotate, scale and offset curve tools.

Copy and move is not included in this, since the QgsAvoidIntersectionsOperation is part of "app", but the copying and moving is done in core code (QgsVectorLayerTools) and I assume you're not supposed to include headers from app in core.

JuhoErvasti committed:
Make move tool remove intersection

JuhoErvasti committed:
Make rotate tool remove intersection

JuhoErvasti committed:
Make scale tool remove intersections

JuhoErvasti committed:
Make offset curve tool remove intersections

JuhoErvasti committed:
Refactor showing error message when intersection could not be removed

troopa81 reviewed (commented):
Thank you for this work! 

You deal with "move tool" but forget to deal with "copy-move tool"

Do you mind adding test for each tool?

troopa81 commented on src/app/qgsmaptoolscalefeature.cpp:
```diff
@@ -368,6 +369,20 @@ void QgsMapToolScaleFeature::applyScaling( double scale )
+  connect( &avoidIntersections, &QgsAvoidIntersectionsOperation::messageEmitted, this, &QgsMapTool::messageEmitted );
+
+  // when removing intersections don't check for intersections with selected features
+  QSet<QgsFeatureId> ignoreFeatureIds;
+  for ( const auto &f : mScaledFeatures )
+  {
+    ignoreFeatureIds.insert( f );
+  }
+
+  QHash<QgsVectorLayer *, QSet<QgsFeatureId> > ignoreFeatures;
+  ignoreFeatures.insert( vlayer, ignoreFeatureIds );
```
same comment as above

JuhoErvasti committed:
Address review

JuhoErvasti commented on src/app/qgsmaptoolmovefeature.cpp:
```diff
@@ -227,6 +242,23 @@ void QgsMapToolMoveFeature::cadCanvasReleaseEvent( QgsMapMouseEvent *e )
+            if ( res.operationResult == Qgis::GeometryOperationResult::InvalidInputGeometryType || geom.isEmpty() )
+            {
+              QString errorMessage = ( geom.isEmpty() ) ?
+                                     tr( "Resulting geometry would be empty" ) :
```
I agree, good point.

JuhoErvasti commented on src/app/qgsmaptoolscalefeature.cpp:
```diff
@@ -378,6 +393,25 @@ void QgsMapToolScaleFeature::applyScaling( double scale )
 
     const QgsFeatureId id = feat.id();
+
+    if ( vlayer->geometryType() == Qgis::GeometryType::Polygon )
```
Aren't all them included in GeometryType::Polygon though? When manually testing the intersection removal does work for all polygon types.

troopa81 replied:
Oh yes indeed! this is the simplified type already not the wkb one, you can forget my comment

JuhoErvasti commented:
@troopa81 Thank you for the review!

I don't mind adding tests for the tools, but it's probably going to take some time.

About the copy-move feature, do you have a suggestion on how to approach it? 

Currently the copy-moving is being done in QgsVectorLayerTools which is core code whereas the QgsAvoidIntersectionsOperation is part of the app code. I'm not too familiar with the code-base, but I assume you're not supposed to include code from the maptools to core.

JuhoErvasti committed:
Add test for avoiding intersections with move tool

JuhoErvasti committed:
Add test for avoid overlap + topo editing with

rotate tool

JuhoErvasti committed:
Add test for avoid overlap + topo editing with

scale tool

troopa81 commented:
> I don't mind adding tests for the tools, but it's probably going to take some time.

I don't mind waiting, the rest of the PR is good now (except maybe the question regarding copy-move) so as soon as there is test for each tool, I'll merge. You can also split this PR in several ones if you want, one for each tool for instance, so you get your work merged progressively as you write the tests.

> About the copy-move feature, do you have a suggestion on how to approach it?

> Currently the copy-moving is being done in QgsVectorLayerTools which is core code whereas the QgsAvoidIntersectionsOperation is part of the app code. I'm not too familiar with the code-base, but I assume you're not supposed to include code from the maptools to core.

No, you're not. Actually QgsMapToolMoveFeature uses a QgsGuiVectorLayerTool which lives in app. So If I were to do this, I think I would override copyMoveFeatures (the method is virtual) add add a virtual protected method avoidIntersection() that would call QgsAvoidIntersectionOperation in QgsGuiVectorLayerTool and would propose another implementation in core without all the Gui stuff. Something like that.

But It would be better to have all this logic in a different PR. I'm a little bothered to have a move tool that behaves differently from the copy-move tool but all the tool are supposed to support avoid intersection so I think the situation would be better after your PR even if the copy-move part is not implemented.

JuhoErvasti commented:
Okay, thank you. I think I'd prefer to add the tests for each tool in this PR and add the copy-move part in a different one. I think I can get the tests done before the end of this week but the copy-move logic might take a bit longer.

troopa81 commented:
> Okay, thank you. I think I'd prefer to add the tests for each tool in this PR and add the copy-move part in a different one. I think I can get the tests done before the end of this week but the copy-move logic might take a bit longer.

:+1:

JuhoErvasti committed:
Add test for offset curve tool

JuhoErvasti committed:
Use cancel() if geom is empty

JuhoErvasti committed:
Fix style

JuhoErvasti committed:
Add test for control modifier

JuhoErvasti committed:
Offset curve: take control modifier into

account when removing intersections

troopa81 commented on src/app/qgsmaptooloffsetcurve.cpp:
```diff
@@ -346,6 +347,30 @@ void QgsMapToolOffsetCurve::applyOffset( double offset, Qt::KeyboardModifiers mo
+
+  connect( &avoidIntersections, &QgsAvoidIntersectionsOperation::messageEmitted, this, &QgsMapTool::messageEmitted );
+
+  const QSet<QgsFeatureId> ignoredFeatures = ( QgsProject::instance()->avoidIntersectionsMode() != Qgis::AvoidIntersectionsMode::AllowIntersections && ( modifiers & Qt::ControlModifier ) ) ?
```
Is `QgsProject::instance()->avoidIntersectionsMode() != Qgis::AvoidIntersectionsMode::AllowIntersections`  really needed because if mode allows intersection the apply() method would do nothing and just return

JuhoErvasti replied:
Ah, no it's not. That was an oversight.

JuhoErvasti committed:
Remove unnecessary check for avoid overlap mode

JuhoErvasti committed:
Change ingroup comment back

troopa81 commented:
@JuhoErvasti Is this PR ready or is there remaining work to do ? all the comments have been addressed?

JuhoErvasti commented:
@troopa81 Yes, this PR should be ready. I've addressed the comments and added tests for each tool.

As for the copy-move implementation; I started preparing a fix and it's mostly working but still needs some development, but that'd be a different PR regardless.

troopa81 reviewed (approved):
Great work, thanks :+1: 

> As for the copy-move implementation; I started preparing a fix and it's mostly working but still needs some development, but that'd be a different PR regardless.

:+1:
